### PR TITLE
Incremental 14+

### PR DESCRIPTION
**Pull request by rgayatri23**
Rest of the incremental unit tests for parallel_reduce, parallel_scan and AtomicViews.

### CODE REVIEW

File: core/unit_test/incremental/Test14_MDRangeReduce.hpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        KOKKOS_LAMBDA(const int i, const int j, value_type& update_value) {
+          update_value += d_data(i, j);
+        },
+        d_resultView);
+
+    // Check correctness.
+    ASSERT_EQ(h_result, d_result);
+
+    // Copy view back to host.
+    Kokkos::deep_copy(view_results, d_resultView);
```
Comment (dalg24) [inline core/unit_test/incremental/Test14_MDRangeReduce.hpp:141]: Why is `view_results` plural.  Avoid declaring it so far away from the place you actually use it.

File: core/unit_test/incremental/Test16_ParallelScan.hpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    View_1D d_data("data", N);
+
+    // Initialize data.
+    Kokkos::parallel_for(
+        N, KOKKOS_LAMBDA(const int i) { d_data(i) = i * 0.5; });
+
+    // Exclusive parallel_scan call.
+    Kokkos::parallel_scan(
+        N, KOKKOS_LAMBDA(const int i, value_type &update_value,
+                         const bool &final) {
```
Comment (dalg24) [inline core/unit_test/incremental/Test16_ParallelScan.hpp:72]: Why do you pass a boolean by const reference?

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  // 1D atomic view
+  using atomic_view =
+      typename Kokkos::View<value_type *, ExecSpace,
+                            Kokkos::MemoryTraits<Kokkos::Atomic> >;
+  using host_atomic_view = typename atomic_view::HostMirror;
+
+  void atomicView() {
+    auto t = time(nullptr);
+    srand(t);  // Use current time as seed for random generator
```
Comment (dalg24) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:75]: This is a bad idea.  We want reproducible runs.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    int *correct_results = new int[num_buckets];
+
+    // Initialize host side histogram arrays
+    for (int i = 0; i < num_buckets; ++i) {
+      h_hist(i)          = 0;
+      correct_results[i] = 0;
+    }
+
+    // Initialize integer input data with random integers
+    for (int i = 0; i < N; ++i) h_data(i) = rand() % 10 * i;
```
Comment (dalg24) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:95]: Prefer `std::uniform_int_distribution`

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp (hunk 3 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test17_CompleteAtomic.hpp]: Using uniform_int_distribution now.

File: core/unit_test/incremental/Test16_ParallelScan.hpp (hunk 2 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test16_ParallelScan.hpp:72]: That is the signature of the parallel_scan call according to the wiki. 
https://github.com/kokkos/kokkos/wiki/ParallelDispatch

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp (hunk 4 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test17_CompleteAtomic.hpp]: Retest this please.

File: core/unit_test/incremental/Test16_ParallelScan.hpp (hunk 2 above)
Comment (dalg24) [reply core/unit_test/incremental/Test16_ParallelScan.hpp:72]: Good catch.  I updated the wiki.

File: core/unit_test/incremental/Test14_MDRangeReduce.hpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+  // Custom Reduction
+  void reduce_custom() {
+    Complex_View_1D d_data("complex array", N);
+    MyComplex result(0.0, 0.0);
+    int sum = 0;
+
+    // Fill data
+    Kokkos::parallel_for(
+        N, KOKKOS_LAMBDA(const int i) {
```
Comment (dalg24) [inline core/unit_test/incremental/Test14_MDRangeReduce.hpp:154]: You probably should not be relying on the default execution space here

File: core/unit_test/incremental/Test16_ParallelScan.hpp (hunk 2 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test16_ParallelScan.hpp]: Great. Modified the code too.

File: core/unit_test/incremental/Test14_MDRangeReduce.hpp (hunk 5 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test14_MDRangeReduce.hpp]: Added the execution space information to parallel_for.

Comment (rgayatri23) [reply core/unit_test/incremental/Test14_MDRangeReduce.hpp]: Retest this please.

**Comment by masterleinad** [comment]
```
[ RUN      ] SERIAL.IncrTest_16_parallelscan
:0: : block: [0,0,0], thread: [0,0,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,1,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,2,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,3,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,4,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,5,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,6,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,7,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,8,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.
:0: : block: [0,0,0], thread: [0,9,0] Assertion `Kokkos::View ERROR: attempt to access inaccessible memory space` failed.

unknown file: Failure
C++ exception with description "cudaFuncSetCacheConfig( cuda_parallel_launch_local_memory<DriverType>, (prefer_shmem ? cudaFuncCachePreferShared : cudaFuncCachePreferL1)) error( cudaErrorAssert): device-side assert triggered /var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:426

Traceback functionality not available
" thrown in the test body.
[  FAILED  ] SERIAL.IncrTest_16_parallelscan (3 ms)
[ RUN      ] SERIAL.incr_17_atomicView
unknown file: Failure

C++ exception with description "cudaDeviceSynchronize() error( cudaErrorAssert): device-side assert triggered /var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:143

Traceback functionality not available
" thrown in the test body.
```

**Comment by rgayatri23** [comment]
Retest this please.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    // Fill host data with integers from the distribution object.
+    for (int i = 0; i < N; ++i) h_data(i) = distribution(generator);
+
+    // Copy data from host to device
+    Kokkos::deep_copy(d_data, h_data);
+    Kokkos::deep_copy(d_hist, h_hist);
+
+    // Update histogram
+    Kokkos::parallel_for(
+        N, KOKKOS_LAMBDA(const int i) { d_hist(d_data(i) % num_buckets)++; });
```
Comment (masterleinad) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:104]: This looks suspicious. You don't specify the execution space so it is running on the host but your accessing data on the device.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+// In this test we sort N integers into num_buckets number of buckets based on
+// their rermainder, i.e., a histogram based on remainder. Since the number of
+// integers is greater than the number of buckets, we use atomic views for the
+// sorted histogram.
+
+namespace Test {
+
+using value_type = int;
+const int N      = 50;
+#define num_buckets 10
```
Comment (masterleinad) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:60]: Any reason to use a preprocessor define here?
A `constexpr` variable should also work.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    // Device and Host views of N number of integers
+    View d_data("deviceData_1D", N);
+    auto h_data = create_mirror_view(d_data);
+
+    // Atomic Device and Host views of histogram
+    atomic_view d_hist("histogram", num_buckets);
+    auto h_hist = create_mirror_view(d_hist);
+
+    // An array to store correct results for verification
+    int *correct_results = new int[num_buckets];
```
Comment (masterleinad) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:87]: You can use `std::array` instead.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+    // Copy the histogram back to host
+    Kokkos::deep_copy(h_hist, d_hist);
+
+    // Validate results
+    for (int i = 0; i < num_buckets; ++i)
+      ASSERT_EQ(correct_results[i], h_hist(i));
+
+    // Free memory allocated for correct results.
+    delete[] correct_results;
```
Comment (masterleinad) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:118]: ...and then you don't need to think about memory management.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+// Incremental test for atomic views.
+// In this test we sort N integers into num_buckets number of buckets based on
+// their rermainder, i.e., a histogram based on remainder. Since the number of
+// integers is greater than the number of buckets, we use atomic views for the
+// sorted histogram.
+
+namespace Test {
+
+using value_type = int;
+const int N      = 50;
```
Comment (masterleinad) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:59]: Not sure if this is large enough. Maybe rather choose 1000 or so?

File: core/unit_test/incremental/Test14_MDRangeReduce.hpp [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+namespace Test {
+using value_type = double;
+const int N      = 10;
+const int M      = 10;
+
+// A structure for complex number.
+struct MyComplex {
```
[the remark this answers is not in the document]
Comment (rgayatri23) [reply core/unit_test/incremental/Test14_MDRangeReduce.hpp:58]: The reason I used my own version of complex was to test the custom reduction for parallel_reduce. 
Hence wanted a structure with everything defined inside.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp (hunk 10 above)
Comment (rgayatri23) [reply core/unit_test/incremental/Test17_CompleteAtomic.hpp]: Yeah made it to 1000.

File: core/unit_test/incremental/Test17_CompleteAtomic.hpp [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    for (int i = 0; i < N; ++i) correct_results[h_data(i) % num_buckets]++;
+
+    // Copy the histogram back to host
+    Kokkos::deep_copy(h_hist, d_hist);
+
+    // Validate results
+    for (int i = 0; i < num_buckets; ++i)
+      ASSERT_EQ(correct_results[i], h_hist(i));
+
+    // Free memory allocated for correct results.
```
Comment (masterleinad) [inline core/unit_test/incremental/Test17_CompleteAtomic.hpp:118]:
```suggestion
```

**Comment by dalg24** [comment]
intermeditery -> intermediary in last commit message

Also they are no changes associated to it.  Rebase gone wrong?

**Comment by masterleinad** [comment]
Just say if you need help with rebasing/squashing.

**Pull request merged**